src/Swarm/TUI/Controller.hs

```diff
@@ -341,31 +348,45 @@ toggleModal mt = do
       newModal <- gets $ flip generateModal mt
       ensurePause
       uiState . uiModal ?= newModal
-    Just _ -> uiState . uiModal .= Nothing >> maybeUnpause
+    Just _ -> uiState . uiModal .= Nothing >> safeAutoUnpause
  where
-  -- these modals do not pause the game
-  runningModals = [RobotsModal]
   -- Set the game to AutoPause if needed
   ensurePause = do
     pause <- use $ gameState . paused
-    unless (pause || mt `elem` runningModals) $ do
+    unless (pause || isRunningModal mt) $ do
       gameState . runStatus .= AutoPause
 
+-- | The running modals do not autopause the game.
+isRunningModal :: ModalType -> Bool
+isRunningModal mt = mt `elem` [RobotsModal, MessagesModal]
+
 handleModalEvent :: V.Event -> EventM Name AppState ()
-handleModalEvent = \case
-  V.EvKey V.KEnter [] -> do
-    mdialog <- preuse $ uiState . uiModal . _Just . modalDialog
-    toggleModal QuitModal
-    case dialogSelection <$> mdialog of
-      Just (Just QuitButton) -> quitGame
-      Just (Just (NextButton scene)) -> startGame scene
-      _ -> return ()
-  ev -> do
-    Brick.zoom (uiState . uiModal . _Just . modalDialog) (handleDialogEvent ev)
-    modal <- preuse $ uiState . uiModal . _Just . modalType
-    case modal of
-      Just _ -> handleInfoPanelEvent modalScroll (VtyEvent ev)
-      _ -> return ()
+handleModalEvent ev = do
+  mt <- preuse $ uiState . uiModal . _Just . modalType
+  let isRunning = maybe False isRunningModal mt
+  case ev of
+    V.EvKey V.KEnter [] -> do
+      mdialog <- preuse $ uiState . uiModal . _Just . modalDialog
+      toggleModal QuitModal
+      case dialogSelection <$> mdialog of
+        Just (Just QuitButton) -> quitGame
+        Just (Just (NextButton scene)) -> startGame scene
+        _ -> return ()
+    -- TODO: copy pasted here from 'handleMainEvent'
```

Comment:
What needs to be done here? Per our new policy, we should either do it or turn it into an issue. :smile:

Reply:
I would like to avoid duplication, but I am not sure if it is feasible.

The modal handler is plugged in higher than the time handlers, so it overrides them.
But maybe they could be conditionally disabled there and moved up? :thinking:

Reply:
Turns out that was pretty easy and it seems to work well. :+1: